binapigen/vppapi/api_schema.go

```diff
@@ -94,4 +97,13 @@ type (
 		StreamMsg string   `json:",omitempty"`
 		Events    []string `json:",omitempty"`
 	}
+
+	Counter struct {
+		Name     string
+		Elements []map[string]string
+	}
+
+	Path struct {
+		List []map[string]string
```

Comment:
This should have more explicit type definition as well. However, I am not totally certain here..

From what I understand and gathered from the current use of the counters & paths in VPP API, it seems to me that paths are a simple list of keys used for stats data related to a specific counter. 

Paths in the `ip.api` file:

```
paths {
  "/err/ip-frag" "ip_frag";
  "/err/mpls-frag" "ip_frag";
  "/err/ip4-mpls-label-disposition-pipe" "ip4";
  "/err/ip4-mpls-label-disposition-uniform" "ip4";
```

look like this in `ip.api.json` file:

```json
    "paths": [
        [
            {
                "path": "/err/ip-frag",
                "counter": "ip_frag"
            },
            {
                "path": "/err/mpls-frag",
                "counter": "ip_frag"
            },
            {
                "path": "/err/ip4-mpls-label-disposition-pipe",
                "counter": "ip4"
            },
            {
                "path": "/err/ip4-mpls-label-disposition-uniform",
                "counter": "ip4"
            },
            {
```

## 1. alternative

For this reason it would make sense to make a map with **counter name** as key and **path list** as value:


```go
type File struct {
    // ...
    
    Paths map[string][]string
}
```

## 2. alternative

or to keep the original order and make it more consistent with the structure of counters:

```go
type File struct {
    // ...
    
    Paths []CounterPaths
}

type CounterPaths struct {
    Name string
    Paths []string
}
```

## 3. alternative

or perhaps another alternative, which uses the assumptions above, we could include the paths in the counters:


```go
type File struct {
    // ...
    
    Counters []Counter
    // no Paths at the file level
}

type Counter struct {
    Name     string
    Elements []CounterElement
    Paths    []string            // <--- instead, put paths here
}
```

This would make all definitions related to a specific counter be at one place.

Reply:
CC @sknat

Reply:
@ondrej-fabry Implemented as the second proposed alternative.
I like the third one the most, but it is not currently applicable as some counter names differ in counters and paths within the same file (`bfd_udp` vs `bfd`, `ikev2` vs `ike`, `mms_clamp` vs `mms-clamp`, etc) and I'd avoid any guessing here.